interpreter/runtimecost.go

```diff
@@ -0,0 +1,185 @@
+	var cost uint64
+	if c.Estimator != nil {
+		callCost := c.Estimator.CallCost(call.OverloadID(), argValues)
+		if callCost != nil {
+			cost += *callCost
+			return cost
+		}
+	}
+	// if user didn't specify, the default way of calculating runtime cost would be used.
+	// if user has their own implementation of ActualCostEstimator, make sure to cover the mapping between overloadId and cost calculation
+	switch call.OverloadID() {
+	// O(n) functions
+	case overloads.StartsWithString, overloads.EndsWithString, overloads.StringToBytes, overloads.BytesToString:
+		cost += uint64(math.Ceil(float64(c.actualSize(argValues[0])) * 0.1))
+	case overloads.InList:
+		// If a list is composed entirely of constant values this is O(1), but we don't account for that here.
+		// We just assume all list containment checks are O(n).
+		cost += c.actualSize(argValues[1])
+	// O(min(m, n)) functions
+	case overloads.LessString, overloads.GreaterString, overloads.LessEqualsString, overloads.GreaterEqualsString,
+		overloads.LessBytes, overloads.GreaterBytes, overloads.LessEqualsBytes, overloads.GreaterEqualsBytes,
+		overloads.Equals, overloads.NotEquals:
+		// When we check the equality of 2 scalar values (e.g. 2 integers, 2 floating-point numbers, 2 booleans etc.),
+		// the CostTracker.actualSize() function by definition returns 1 for each operand, resulting in an overall cost
+		// of 1.
+		lhsCost := c.actualSize(argValues[0])
+		rhsCost := c.actualSize(argValues[1])
+		if lhsCost > rhsCost {
+			cost += rhsCost
+		} else {
+			cost += lhsCost
+		}
+	// O(m+n) functions
+	case overloads.AddString, overloads.AddBytes, overloads.AddList:
+		// In the worst case scenario, we would need to reallocate a new backing store and copy both operands over.
+		cost += uint64(math.Ceil(float64(c.actualSize(argValues[0])+c.actualSize(argValues[1])) * 0.1))
+	// O(nm) functions
+	case overloads.MatchesString:
+		// https://swtch.com/~rsc/regexp/regexp1.html applies to RE2 implementation supported by CEL
+		strCost := uint64(math.Ceil(float64(c.actualSize(argValues[0])) * 0.1))
+		// We don't know how many expressions are in the regex, just the string length (a huge
+		// improvement here would be to somehow get a count the number of expressions in the regex or
+		// how many states are in the regex state machine and use that to measure regex cost).
+		// For now, we're making a guess that each expression in a regex is typically at least 4 chars
+		// in length.
+		regexCost := uint64(math.Ceil(float64(c.actualSize(argValues[1])) * 0.25))
+		cost += strCost * regexCost
+	case overloads.ContainsString:
+		strCost := uint64(math.Ceil(float64(c.actualSize(argValues[0])) * 0.1))
+		substrCost := uint64(math.Ceil(float64(c.actualSize(argValues[1])) * 0.1))
+		cost += strCost * substrCost
+	default:
+		// The following operations are assumed to have O(1) complexity.
+		// 1. Concatenation of 2 lists: see the implementation of the concatList type.
```

Comment:
For the concatenation of two lists this cost is true, but there's an O(c) multiplier on indexed lookups for concatenated lists where `c` is the number of concatenations performed. In most cases its simplest to consider concatenation as the creation of a new list containing all of the elements of each list. The one exception to this rule is when the right hand side argument only contains a single entry, in which case it's likely an append within a comprehension and should only cost 1. 

Kind of weird and tricky, but hopefully the above guidelines give you a better idea.

Reply:
I follow the optimization. Thanks for explaining.  I've updated the cost to match (I don't track the index lookup cost-- it sounds like worst case becomes O(n) lengh t of the list, but I'm willing to wave my hands on that one right now).